rubix-presto/src/main/java/com/qubole/rubix/presto/PrestoClusterManager.java

```diff
@@ -12,149 +12,170 @@
+                                return ImmutableList.of(InetAddress.getLocalHost().getHostName());
+                            }
+
+                            if (failedNodes.isEmpty()) {
+                                failedNodes = ImmutableList.of();
+                                //return ImmutableList.of(InetAddress.getLocalHost().getHostName());
+                            }
+
+                            // keep only the healthy nodes
+                            allNodes.removeAll(failedNodes);
+
+                            Set<String> hosts = new HashSet<String>();
+
+                            for (Stats node : allNodes) {
+                                hosts.add(node.getUri().getHost());
+                            }
+                            if (hosts.isEmpty()) {
+                                // case of master only cluster
+                                hosts.add(InetAddress.getLocalHost().getHostName());
+                            }
+                            List<String> hostList = Lists.newArrayList(hosts.toArray(new String[0]));
+                            Collections.sort(hostList);
+                            return hostList;
                         }
-                        else {
-                            failedNodes = (List<Stats>) failedNodesResponse.getValue();
+                        catch (IOException e) {
+                            throw Throwables.propagate(e);
                         }
                     }
-
-                    // keep only the healthy nodes
-                    allNodes.removeAll(failedNodes);
-
-                    Set<String> hosts = new HashSet<String>();
-
-                    for (Stats node : allNodes) {
-                        hosts.add(node.getUri().getHost());
-                        log.debug(String.format("Node: %s", node.getUri()));
-                    }
-
-                    if (hosts.isEmpty()) {
-                        // case of master only cluster
-                        hosts.add(InetAddress.getLocalHost().getHostName());
-                    }
-
-                    List<String> hostList = Lists.newArrayList(hosts.toArray(new String[0]));
-                    Collections.sort(hostList);
-                    return hostList;
-                }
-                catch (InterruptedException | ExecutionException | URISyntaxException | UnknownHostException e) {
-                    throw Throwables.propagate(e);
-                }
-            }
-        }, 10, TimeUnit.SECONDS);
+                }, executor));
     }
 
     @Override
     public boolean isMaster()
     {
         // issue get on nodesSupplier to ensure that isMaster is set correctly
-        nodesSupplier.get();
+        //nodesSupplier.get();
+        try {
+            nodesCache.get("nodeList");
+        }
+        catch (ExecutionException e) {
+            e.printStackTrace();
```

Comment:
isnt this wrong? When we hit an exception, we should let the caller know about it because we dont know if isMaster set is correct or not.